### PR TITLE
Improve collective ops on numeric input

### PR DESCRIPTION
**Pull request by sadra-barikbin**
Description:
When `ComputationModel._collective_op_dtype` is `None` and the input to a
collective op is a float number, the constructed intermediate tensor is of
dtype `torch.float32` rather than `torch.double`, hence we would have a loss
in precision.

Check list:

- [ ] New tests are added (if a new feature is added)
- [ ] New doc strings: description and/or example code are in RST format
- [ ] Documentation is updated (if required)

### CODE REVIEW

File: ignite/distributed/comp_models/base.py [hunk 1]
```diff
@@ -187,7 +187,11 @@ def _collective_op(
         device = self.device()
         if isinstance(tensor, (Number, float)):
             tensor_to_number = True
-            tensor = torch.tensor(tensor, device=device, dtype=self._collective_op_dtype)
+            if self._collective_op_dtype is None:
+                dtype = torch.double if isinstance(tensor, float) else torch.int64
+            else:
+                dtype = self._collective_op_dtype
```
Comment (vfdev-5) [inline ignite/distributed/comp_models/base.py:193]:
```suggestion
            dtype = self._collective_op_dtype
            if dtype is None and isinstance(tensor, float):
                dtype = torch.double
```

**Comment by vfdev-5** [comment]
@sadra-barikbin where do you spot such precision loss ?

**Comment by sadra-barikbin** [comment]
> @sadra-barikbin where do you spot such precision loss ?

When I attempted to remove the need to use `pytest.approx` in `_test_distrib_on_output`, I ended up here.
After this PR I saw that there was no need to `approx` and `==` worked even without this change, however this change is correct.

**Review by vfdev-5 — APPROVED** [review]
LGTM, thanks for the PR @sadra-barikbin
